torchtext/data/metrics.py

```diff
@@ -0,0 +1,93 @@
+import math
+import collections
+from torchtext.data.utils import ngrams_iterator
+
+
+def bleu_score(candidate_corpus, references_corpus, max_n=4, weights=[0.25] * 4):
+    """Computes the BLEU score between a candidate translation corpus and a references
+    translation corpus. Based on https://www.aclweb.org/anthology/P02-1040.pdf
+
+    Arguments:
+        candidate_corpus: a list of candidate translations. Each translation is a list of
+            tokens
+        references_corpus: a list of lists of reference translations. Each translation
+            is a list of tokens
+        max_n: the maximum n-gram we want to use. E.g. if max_n=3, we will use unigrams,
+            bigrams and trigrams
+        weights: weights used for each n-gram category (uniform by default)
+
+    Examples:
+        >>> candidate_corpus = [['I', 'ate', 'the', 'apple'], ['I', 'did']]
+        >>> references_corpus = [[['I', 'ate', 'it'], ['I', 'ate', 'apples']],
+                [['I', 'did']]]
+        >>> bleu_score(candidate_corpus, references_corpus)
+        >>> 0.7598356856515925
+    """
+
+    def _compute_ngram_counter(tokens, max_n):
+        """ Create a Counter with a count of unique n-grams in the tokens list
+
+        Arguments:
+            tokens: a list of tokens (typically a string split on whitespaces)
+            max_n: the maximum order of n-gram wanted
+
+        Outputs:
+            output: a collections.Counter object with the unique n-grams and their
+                associated count
+
+        Examples:
+            >>> from torchtext.data.functional import _compute_ngram_counter
+            >>> tokens = ['me', 'me', 'you']
+            >>> _compute_ngram_counter(tokens, 2)
+                Counter({('me',): 2,
+                 ('you',): 1,
+                 ('me', 'me'): 1,
+                 ('me', 'you'): 1,
+                 ('me', 'me', 'you'): 1})
+        """
+        assert max_n > 0
+
+        ngrams = [tuple(x.split(' ')) for x in ngrams_iterator(tokens, max_n)]
+        ngrams_counter = collections.Counter(ngrams)
+
+        return ngrams_counter
+
+    assert max_n == len(weights), 'Length of the "weights" list has be equal to max_n'
+
+    clipped_counts = [0.0] * max_n
+    total_counts = [0.0] * max_n
+
+    candidate_len = 0.0
+    refs_len = 0.0
+
+    for (candidate, refs) in zip(candidate_corpus, references_corpus):
+        candidate_len += len(candidate)
+
+        # Get the length of the reference that's closest in length to the candidate
+        refs_len_list = [float(len(ref)) for ref in refs]
+        refs_len = min(refs_len_list, key=lambda x: abs(len(candidate) - x))
+
+        reference_counters = collections.Counter()
+        for ref in refs:
+            reference_counters = reference_counters | _compute_ngram_counter(ref, max_n)
```

Comment:
We carry around the count for tokens we will discard as soon as we see the candidate_counter. I wonder if candidate_counter could be used as an online filter when accumulating reference_counters, so we don't need to maintain that memory. It's also regrettable that we have to create a new collection for every reference. I'm sure most of the time that's fine, because many datasets will only have one target reference. It could be worth writing around that special case via
```
reference_counters = _compute_ngram_counter(refs[0], max_n)
for ref in refs[1:]:
   ...
```